### DISCUSSION TITLE
extracting velocity data from a simulation

### DISCUSSION THREAD

**Discussion by valerubianol** [pyswmm/pyswmm#537] [Q&A] [answered]
Hello everyone, as a student I am trying to do a project on sewage. That's why I'm using pyswmm but i`m having trouble extracting velocity values ​​from my simulation. If anyone can help me I would appreciate it because honestly, programming has been a challenge for me. 

this is what im doing,

from pyswmm import Simulation, Links
import pandas as pd

# Ruta al archivo .inp
inp_file = 'D:\\escritorio\\TESISfinal\\programacion interfaz\\interfaz\\modelo_tiempo_seco.inp'

# Crear listas para almacenar los resultados
times = []
depths = []
velocities = []
flows = []

# Crear una simulación
with Simulation(inp_file) as sim:
    # Acceder a todos los enlaces (links) en el modelo
    links = Links(sim)
    
    # Iterar sobre cada paso de tiempo de la simulación
    for step in sim:
        current_time = sim.current_time
        for link in links:
            times.append(current_time)
            depths.append(link.depth)
            velocities.append(link.velocity)
            flows.append(link.flow)

# Crear un DataFrame con los resultados
results = pd.DataFrame({
    'Time': times,
    'Depth': depths,
    'Velocity': velocities,
    'Flow': flows
})

# Guardar el DataFrame en un archivo CSV
results.to_csv('simulation_results.csv', index=False)

print("Simulación completada y resultados guardados en 'simulation_results.csv'.")

**Comment by bemcdonnell** [comment]
@valerubianol,

https://pyswmm.github.io/pyswmm/reference/api/pyswmm.links.Link.html# `link.velocity` is not an option.  However, you can get the average velocity between the upstream and downstream end of the pipe by averaging the flow areas. 

V = Q / ( (A_upstream + A_downstream ) / 2 )

In pyswmm this would look like:

```python
from pyswmm import Simulation, Links

with Simulation("path.inp") as sim:
    link1 = Links(sim)['Link_name']

    for step in sim:
        area_us = link1.ups_xsection_area
        area_ds = link1.ds_xsection_area
        q = link1.flow

        V = q / ( (area_us + area_ds) / 2.0 )
```

**Reply by valerubianol** [reply]
Thank you so much for your answer, but when I tried to do it that way the result is zero 😔 could you please give me a different form of extracting the velocity from the links in the simulation

**Reply by valerubianol** [reply]
ZeroDivisionError: float division by zero

**Reply by bemcdonnell** [reply]
@valerubianol, so if the denominators sum to zero for a time step, maybe consider adding a check to make sure it doesn’t try to divide by zero? To me that means the velocity is zero. Does that make sense?

**Comment by alejorojas95** [comment]
I have the same problems, I need a more efficient speed calculation and it gives me the same problem with the 
inp inp_file = 'inp/swmm_example.inp'  and when doing the following: 


```python
from pyswmm import Simulation, Links , Nodes
import pprint
import math
import numpy as np
# Ruta al archivo .inp
inp_file = 'inp/swmm_example.inp'

with Simulation(inp_file) as sim:
    link1 = Links(sim)['C1']
 
    for step in sim:
        area_us = link1.ups_xsection_area
        area_ds = link1.ds_xsection_area
        q = link1.flow
        print("areas: ", area_ds, area_us )
        #V = q / ( (area_us + area_ds) / 2.0 )

```


the print is 
areas: 0.0 0.0
areas: 0.0 0.0

?
![velocity](https://github.com/pyswmm/pyswmm/assets/57581430/c76d48d5-9d9e-4939-a1a3-2e005d6dc26b)

**Reply by bemcdonnell** [reply]
@alejorojas95 those values you screenshot are from the report file which are summaries generated from the whole simulation.  @valerubianol is trying to calculate the discrete velocities for the simulation time steps

**Reply by alejorojas95** [reply]
I appreciate your response, it gave more clarity to my doubts regarding the time intervals, yes sir, and how it would be to obtain those speeds resulting from each link, thank you very much.

``

**Reply by bemcdonnell** [reply]
If it was me, I’d probably sample from some of the links rather that calculate all link velocity results.  It’s a heavy calculation for pyswmm since it has to fetch all those values for each link.  Are you trying to run this in a real time environment or is this a planning application? 


@karosc, is talking about adding a velocity calculation to the API in the release of SWMM-5.3.  @alejorojas95 i bet he’d love the help!

**Reply by alejorojas95** [reply]
> Si fuera yo, probablemente tomaría muestras de algunos de los enlaces en lugar de calcular todos los resultados de la velocidad del enlace. Es un cálculo pesado para pyswmm ya que tiene que recuperar todos esos valores para cada enlace. ¿Está intentando ejecutar esto en un entorno de tiempo real o es una aplicación de planificación?
> 
> @karosc, habla de agregar un cálculo de velocidad a la API en la versión de SWMM-5.3. @alejorojas95¡Apuesto a que le encantaría recibir ayuda!

The idea is to extract some data from the simulation, for example the area, diameter, froude, inundaicone to validate and comply with regulatory conditions, friend. I appreciate your response

**Comment by bemcdonnell** [comment, marked as answer]
@valerubianol and @alejorojas95 

```Python
with Simulation(inp_file) as sim:
    link1 = Links(sim)['C1']
 
    for step in sim:
        area_us = link1.ups_xsection_area
        area_ds = link1.ds_xsection_area
        q = link1.flow
        if  area_ds + area_us != 0:
            V = q / ( (area_us + area_ds) / 2.0 )
        else:
            V = 0
```
